src/tribler/core/components/session.py

```diff
@@ -51,7 +51,20 @@ async def __aexit__(self, *_):
         await self.shutdown()
 
     def get_instance(self, comp_cls: Type[T]) -> Optional[T]:
-        return self.components.get(comp_cls)
+        # try to find a direct match
+        if direct_match := self.components.get(comp_cls):
+            return direct_match
+
+        # try to find a subclass match
+        candidates = {c for c in self.components if issubclass(c, comp_cls)}
+
+        if not candidates:
+            return None
+        if len(candidates) >= 2:
+            raise KeyError(f'Found more that two subclasses for the class {comp_cls}')
```

Comment:
**nitpick:** It is better to use a new exception class (something like `MultipleComponentsFound`) inherited from `ComponentError`.

With the KeyError exception, it sometimes becomes possible that the method is expected to raise KeyError in some situations (like, multiple components found), but also it raises the same KeyError exception for a totally unrelated reason (incorrect access to some list or dict), and it goes unnoticed.

Also, the standard [KeyError](https://docs.python.org/3/library/exceptions.html#KeyError) exception has the following description: "Raised when a mapping (dictionary) key is not found in the set of existing keys." There we have a different situation when multiple objects are found, so using KeyError looks incorrect.

Reply:
Thank you for this suggestion. It has been implemented.